spring-boot/proxy/src/test/java/eu/coatrack/proxy/security/local_api_key_manager_tests/GatewayModeUpdateLoggingTest.java

```diff
@@ -0,0 +1,151 @@
+        switchToOfflineMode();
+
+        assertEquals(0, logEventStorage.getNumberOfLogEventsContaining(onlineModeMessage));
+        assertEquals(0, logEventStorage.getNumberOfLogEventsContaining(offlineModeMessage));
+    }
+
+    @Test
+    public void ON_OFF() throws ApiKeyFetchingFailedException {
+        switchToOnlineMode();
+        switchToOfflineMode();
+
+        assertEquals(1, logEventStorage.getNumberOfLogEventsContaining(onlineModeMessage));
+        assertEquals(1, logEventStorage.getNumberOfLogEventsContaining(offlineModeMessage));
+    }
+
+    @Test
+    public void OFF_ON() throws ApiKeyFetchingFailedException {
+        switchToOfflineMode();
+        switchToOnlineMode();
+
+        assertEquals(1, logEventStorage.getNumberOfLogEventsContaining(onlineModeMessage));
+        assertEquals(0, logEventStorage.getNumberOfLogEventsContaining(offlineModeMessage));
+    }
+
+    @Test
+    public void ON_ON() throws ApiKeyFetchingFailedException {
+        switchToOnlineMode();
+        switchToOnlineMode();
+
+        assertEquals(1, logEventStorage.getNumberOfLogEventsContaining(onlineModeMessage));
+        assertEquals(0, logEventStorage.getNumberOfLogEventsContaining(offlineModeMessage));
+    }
+
+    @Test
+    public void OFF_OFF() throws ApiKeyFetchingFailedException {
+        switchToOfflineMode();
+        switchToOfflineMode();
+
+        assertEquals(0, logEventStorage.getNumberOfLogEventsContaining(onlineModeMessage));
+        assertEquals(0, logEventStorage.getNumberOfLogEventsContaining(offlineModeMessage));
+    }
+
+    @Test
+    public void localApiKeySearchShouldNotTriggerSwitchToOnlineMode() {
+        localApiKeyManager.getApiKeyEntityByApiKeyValue(apiKey.getKeyValue());
+
+        assertEquals(0, logEventStorage.getNumberOfLogEventsContaining(onlineModeMessage));
+        assertEquals(0, logEventStorage.getNumberOfLogEventsContaining(offlineModeMessage));
+    }
+
+    @Test
+    public void localApiKeySearchShouldTriggerSwitchFromOnlineToOfflineMode() throws ApiKeyFetchingFailedException {
+        switchToOnlineMode();
+        localApiKeyManager.getApiKeyEntityByApiKeyValue(apiKey.getKeyValue());
+
+        assertEquals(1, logEventStorage.getNumberOfLogEventsContaining(onlineModeMessage));
+        assertEquals(1, logEventStorage.getNumberOfLogEventsContaining(offlineModeMessage));
+    }
+
+    public void switchToOnlineMode() throws ApiKeyFetchingFailedException {
+        reset(apiKeyFetcherMock);
+        when(apiKeyFetcherMock.requestLatestApiKeyListFromAdmin()).thenReturn(apiKeyList);
+        localApiKeyManager.updateLocalApiKeyList();
+    }
+
+    public void switchToOfflineMode() throws ApiKeyFetchingFailedException {
+        reset(apiKeyFetcherMock);
+        when(apiKeyFetcherMock.requestLatestApiKeyListFromAdmin()).thenThrow(new ApiKeyFetchingFailedException("test"));
+        localApiKeyManager.updateLocalApiKeyList();
```

Comment:
The relationship between "calling the update method" and "testing the mode logging" is only indirect. So probably this class does not only test the logging, but if the "mode switch" is done correctly?

Reply:
This is correct. I implemented it this way to not destroy the encapsulation. I changed the name of that method to ```letLocalApiKeyManagerSwitchToOfflineMode``` for better readability. 

Do you have further suggestions?

Reply:
I added the comment for further explanation as you suggested. This issue is done.